dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/TrackerObjectsMapper.java

```diff
@@ -0,0 +1,348 @@
+    }
+
+    dbTrackedEntity.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
+    dbTrackedEntity.setLastUpdated(now);
+    dbTrackedEntity.setPotentialDuplicate(trackedEntity.isPotentialDuplicate());
+    dbTrackedEntity.setCreatedAtClient(DateUtils.fromInstant(trackedEntity.getCreatedAtClient()));
+    dbTrackedEntity.setLastUpdatedAtClient(
+        DateUtils.fromInstant(trackedEntity.getUpdatedAtClient()));
+    dbTrackedEntity.setOrganisationUnit(organisationUnit);
+    dbTrackedEntity.setTrackedEntityType(trackedEntityType);
+    dbTrackedEntity.setInactive(trackedEntity.isInactive());
+    dbTrackedEntity.setGeometry(trackedEntity.getGeometry());
+
+    return dbTrackedEntity;
+  }
+
+  public static @Nonnull Enrollment map(
+      @Nonnull TrackerPreheat preheat,
+      @Nonnull org.hisp.dhis.tracker.imports.domain.Enrollment enrollment,
+      @Nonnull UserDetails user) {
+    Enrollment dbEnrollment = preheat.getEnrollment(enrollment.getEnrollment());
+    OrganisationUnit organisationUnit = preheat.getOrganisationUnit(enrollment.getOrgUnit());
+
+    Program program = preheat.getProgram(enrollment.getProgram());
+
+    TrackedEntity trackedEntity = preheat.getTrackedEntity(enrollment.getTrackedEntity());
+
+    Date now = new Date();
+
+    if (dbEnrollment == null) {
+      dbEnrollment = new Enrollment();
+      dbEnrollment.setUid(enrollment.getEnrollment());
+      dbEnrollment.setCreated(now);
+      dbEnrollment.setStoredBy(enrollment.getStoredBy());
+      dbEnrollment.setCreatedByUserInfo(UserInfoSnapshot.from(user));
+    }
+
+    dbEnrollment.setLastUpdated(now);
+    dbEnrollment.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
+    dbEnrollment.setCreatedAtClient(DateUtils.fromInstant(enrollment.getCreatedAtClient()));
+    dbEnrollment.setLastUpdatedAtClient(DateUtils.fromInstant(enrollment.getUpdatedAtClient()));
+
+    Date enrollmentDate = DateUtils.fromInstant(enrollment.getEnrolledAt());
+    Date occurredDate = DateUtils.fromInstant(enrollment.getOccurredAt());
+
+    dbEnrollment.setEnrollmentDate(enrollmentDate);
+    dbEnrollment.setOccurredDate(occurredDate != null ? occurredDate : enrollmentDate);
+    dbEnrollment.setOrganisationUnit(organisationUnit);
+    dbEnrollment.setProgram(program);
+    dbEnrollment.setTrackedEntity(trackedEntity);
+    dbEnrollment.setFollowup(enrollment.isFollowUp());
+    dbEnrollment.setGeometry(enrollment.getGeometry());
+
+    EnrollmentStatus previousStatus = dbEnrollment.getStatus();
+    dbEnrollment.setStatus(enrollment.getStatus());
+
+    if (previousStatus != dbEnrollment.getStatus()) {
+      switch (dbEnrollment.getStatus()) {
+        case ACTIVE -> {
+          dbEnrollment.setCompletedDate(null);
+          dbEnrollment.setCompletedBy(null);
+        }
+        case COMPLETED -> {
+          dbEnrollment.setCompletedDate(now);
+          dbEnrollment.setCompletedBy(user.getUsername());
+        }
+        case CANCELLED -> {
```

Comment:
unrelated to this PR, but I've seen that when the enrollment is `CANCELLED`, the related event(s) are not modified, so it could happen that an event is `ACTIVE` while the parent enrollment is `CANCELLED`, does that make sense?

Reply:
Same as before, I just didn't change the logic, but in https://dhis2.atlassian.net/browse/DHIS2-18224 we should try to understand if it makes sense or not.